src/hotspot/share/services/management.cpp

```diff
@@ -2098,17 +2094,32 @@ jlong Management::ticks_to_ms(jlong ticks) {
 
 // Gets the amount of memory allocated on the Java heap since JVM launch.
 JVM_ENTRY(jlong, jmm_GetTotalThreadAllocatedMemory(JNIEnv *env))
-    // The result may be "too small" because the threads iterator takes a
-    // snapshot and new threads may be created after result is initialized.
-    // There's no check for terminated threads in the loop because their
-    // allocated byte counts aren't included in exited_allocated_bytes
-    // when result is initialized. The iterator is created and initialized
-    // before the call to exited_allocated_bytes to ensure we don't miss
-    // accounting for threads that are just about to terminate.
+    // A thread increments exited_allocated_bytes after being removed from
+    // the threads list (in ThreadService::remove_thread), which means there's
+    // a race between threads that exit during the loop and reading
+    // exited_allocated_bytes. If result is initialized with exited_allocated_bytes,
+    // the final result may be "too small" because a thread might be removed
+    // from the list before the loop gets to it and thus not be counted. If,
+    // on the other hand, exited_allocated_bytes is added after the loop,
+    // the final result might be "too large" because a thread might be counted
+    // twice, once in the loop and agsin in exited_allocated_bytes if it's
+    // removed from the list after it's encountered in the loop but before
+    // adding exited_allocated_bytes.
```

Comment:
This is not an accurate description.

Once we have the TLH we have a fixed set of threads and none of them can update exited_allocated_bytes. If we read exited_allocated_bytes while the TLH is held then the count can be short for two reasons:
1. Newly started threads
2. Terminating threads that already escaped the TLH but which hadn't updated exited_allocated_bytes by the time we read it.
AFAICS with the current logic there is no possibility of double-accounting.

Reply:
Sorry I thought you had made the change to only update exited_allocated_bytes after the thread removed itself from the threads-list. It really makes things so much simpler.

Reply:
I made that change. The comment now reads:
    // A thread increments exited_allocated_bytes in ThreadService::remove_thread                                                                                                                                                                                                                                                                                           
    // only after it removes itself from the threads list, and once a TLH is                                                                                                                                                                                                                                                                                                
    // created, no thread it references can remove itself from the threads                                                                                                                                                                                                                                                                                                  
    // list, so none can update exited_allocated_bytes. We therefore initialize                                                                                                                                                                                                                                                                                             
    // result with exited_allocated_bytes after after we create the TLH so that                                                                                                                                                                                                                                                                                             
    // the final result can only be short due to (1) threads that start after                                                                                                                                                                                                                                                                                               
    // the TLH is created, or (2) terminating threads that escape TLH creation                                                                                                                                                                                                                                                                                              
    // and don't update exited_allocated_bytes before we initialize result.